src/services/codefixes/inferFromUsage.ts

```diff
@@ -414,27 +414,33 @@ namespace ts.codefix {
                 inferTypeFromContext(reference, checker, usageContext);
             }
             const isConstructor = declaration.kind === SyntaxKind.Constructor;
-            const callContexts = isConstructor ? usageContext.constructContexts : usageContext.callContexts;
-            return callContexts && declaration.parameters.map((parameter, parameterIndex): ParameterInference => {
-                const types: Type[] = [];
+            const isJS = isInJSFile(declaration);
+            const callContexts = isJS ? [...usageContext.constructContexts || [], ...usageContext.callContexts || []] :
+                isConstructor ? usageContext.constructContexts :
+                usageContext.callContexts;
+            return declaration.parameters.map((parameter, parameterIndex): ParameterInference => {
+                const types = [];
                 const isRest = isRestParameter(parameter);
                 let isOptional = false;
-                for (const callContext of callContexts) {
-                    if (callContext.argumentTypes.length <= parameterIndex) {
-                        isOptional = isInJSFile(declaration);
-                        types.push(checker.getUndefinedType());
-                    }
-                    else if (isRest) {
-                        for (let i = parameterIndex; i < callContext.argumentTypes.length; i++) {
-                            types.push(checker.getBaseTypeOfLiteralType(callContext.argumentTypes[i]));
+                if (callContexts) {
+                    for (const callContext of callContexts) {
+                        if (callContext.argumentTypes.length <= parameterIndex) {
+                            isOptional = isJS;

```

Comment:
Similarly, does this need to depend on whether we're in JS?

Reply:
I thought about this one, but it's a bigger change -- right now the codefix only adds annotations, it doesn't add other punctuation around parameters. I'll open a bug to track this, though, because you are right, it should work (and it's probably not too hard).

Reply:
Ref:  #28377